### ISSUE TITLE
Revisit pre-prediction

### ISSUE THREAD

**Issue by cBournhonesque** [cBournhonesque/lightyear#66]
Consider the following scenario:
- client pre-spawns bullet
- server must confirm if it's valid or not, and get authority over the bullet

Currently the way to do this would be:
- spawn the bullet on client-side with a `ShouldBePredicted` component, and replicates that to the server.
- server receives the bullet, creates a new one.
- if the bullet is valid, server adds `Replicate` to replicate it back to clients; or deletes it if it's not valid (to replicate the deletion).

This means that the bullet-spawning logic (if client pressed Fire) is very different on client and on server.
Ideally we would like the spawning logic to be similar.

Here is what unity does: https://docs.unity3d.com/Packages/com.unity.netcode@1.2/manual/ghost-spawning.html
- client spawns the bullet on input Fire and indicates that this is pre-Predicted
- server spawns the bullet on input Fire and replicates to client
- when client receives an entity, it checks if it matches any client pre-predicted entities (by comparing archetypes and spawn ticks). If so, it doesn't create a new one but instead re-uses the one it had!

Basically right now we support:
- normal spawning
- spawning player-controlled predicted entity, similar to unity

```
Predicted spawning for the client predicted player object: The object is predicted so the input handling applies immediately. Therefore, it doesn't need to be delay spawned. When the snapshot data for this object arrives, the update system applies the data directly to the object and then plays back the local inputs which have happened since that time, and corrects mistakes in the prediction.
```

but we don't support the 3rd category:
```
Predicted spawning for player spawned objects: This usually applies to objects that the player spawns, like in-game bullets or rockets that the player fires.
```

**Comment by cBournhonesque** [comment]
This is now supported with https://github.com/cBournhonesque/lightyear/pull/91
(Enables pre-spawning player objects)
